Transient Fault Handling

dylan-asos opened the pull request:
This PR implements retry behaviour for transient faults when using HttpClient to send the request. No new dependencies in this PR as per https://github.com/sendgrid/sendgrid-csharp/pull/497#issuecomment-320076981

Default behaviour of retry logic is off. The consumer of the library must explicitly enable this using the ReliabilitySettings and the relevant ctor.

Retry logic implemented as a delegating handler in the HttpClient handlers pipeline, with a simple wait-and-retry (no exponential backoff). Retry behaviour expects to handle

* Timeouts when sending the request that might result in a TaskCanceledException. 
* Http Status codes within the 500 range that indicate a server error. 
* Any other HttpRequestException that might be thrown during the operation.

* NB: Does not handle anything in the 400 range including 429 (where a circuit breaker might be more appropriate)

New object ReliabilitySettings defines retry count and retry interval. Setting retry count > 0 enables the behaviour.

New object SendGridClientOptions defines the settings to use with SendGridClient, and new ctor allows this object,

public SendGridClient(SendGridClientOptions options)

Testing: The retry logic is tested independently of the main Integration test class in new class RetryDelegatingHandlerTests.cs - with a couple of integration tests to verify the collaboration.

thinkingserious commented:
Thanks @dylan-asos! This has been added to our backlog for review.

@Jericho @wildanr @Niladri24dutta @JohnGalt1717 @xt0rted thoughts?

xt0rted commented on src/SendGrid/Reliability/RetryDelegatingHandler.cs:
```diff
@@ -0,0 +1,87 @@
+                    }
+
+                    // ReSharper disable once MethodSupportsCancellation, cancel will be indicated on the token
+                    await Task.Delay(this.settings.RetryInterval);
```
Is it safe for calls to `Task.Delay()` to exclude `.ConfigureAwait(false)`?

thinkingserious replied:
Thanks for jumping in and helping @xt0rted!

Stephen Cleary says you should include it as a stylistic preference: https://stackoverflow.com/questions/43037782/should-i-use-configureawaitfalse-on-my-task-delay-inside-my-task-run-action

dylan-asos replied:
OK sure, I'll add that

Jericho commented:
My main question is: why do you consider `HTTP 500` indication of transient errors? In my experience, errors in the 500 range are indication of server-side errors and retrying the exact same request typically yields the same result. Is it different in SendGrid's API?

As far as I know, the only return code worth retrying (per [SendGrid documentation](https://sendgrid.api-docs.io/v3.0/how-to-use-the-sendgrid-v3-api/api-rate-limits#rate-limit-response-header)) is `HTTP 428`.

Jericho commented on src/SendGrid/SendGridClient.cs:
```diff
@@ -76,6 +82,35 @@ public SendGridClient(IWebProxy webProxy, string apiKey, string host = null, Dic
+            }
+
+            this.options = options;
+            client = (httpClient == null) ? CreateHttpClientWithRetryHandler() : httpClient;
```
When `httpClient` is not null, the retry settings are ignored. Is that intentional?

dylan-asos replied:
Yes, there are integration tests that use that and pass their own handler e.g. TestTakesHttpClientFactoryAsConstructorArgumentAndUsesItInHttpCalls. That preserves existing logic as per https://github.com/sendgrid/sendgrid-csharp/blob/f4092b7ac50353dbe51f0ac70f50ab69f1b61034/src/SendGrid/SendGridClient.cs#L88

Jericho replied:
ok. In that case shouldn't you inform the caller that retry options will be ignored? Maybe by throwing an exception for example. Otherwise someone could pass their own `httpClient` and also expect retry logic.

dylan-asos replied:
Documentation states that you must use the construct that takes the options object for retries - https://github.com/dylan-asos/sendgrid-csharp/blob/d17205d5ffc57d1acae1a747be1b89957d7fee1f/USE_CASES.md#transient_faults.

Jericho replied:
Consider this scenario:
```
var options = new SendGridClientOptions
{
    ApiKey = "Your-Api-Key",
    ReliabilitySettings = {RetryCount = 5, RetryInterval = TimeSpan.FromSeconds(1)}
};

var client = new SendGridClient(myHttpClient, options);
```
In this scenario, I am providing my own http client and also I am expecting retries but my retry settings are silently ignored.

dylan-asos replied:
Yeah fair enough, that ctor could just be internal.

Jericho replied:
I never really investigated but isn't there a way to append your delegating handler to an existing http client? If so, it would allow you to accept a custom http client and also enable retry scenarios.

dylan-asos replied:
The HttpMessageHandler belongs to the HttpClient base class HttpMessageInvoker, you only get to set it at construct time. You can chain up what you want here, but you have to do it through the HttpClient ctor.

Jericho replied:
ok, that's too bad!

thinkingserious commented:
Thanks for the great feedback @Jericho, much appreciated :)

Jericho commented:
@thinkingserious Has SendGrid documented that `HTTP 500` responses are safe to retry? I couldn't find any mention of it in the documentation. As I mentioned before, the only documented retry scenario I could find is `HTTP 428`.

thinkingserious commented:
@Jericho,

I think you are correct regarding `HTTP 429`.

dylan-asos commented:
@thinkingserious - will take your guidance here re 500 errors. Working with various apis, we've generally found these to be worth retrying, particularly 500/503. However, for this initial stage, we could limit this to timeouts and general exceptions that occur at the client, e.g. as per https://github.com/sendgrid/sendgrid-csharp/issues/505.

thinkingserious commented:
@dylan-asos,

I am checking with our team right now. Thanks!

dylan-asos commented:
Some additional commits based on discussion here - found that the ctor with the webproxy wasn't covered by any tests, so have added those. 

I'm done for the weekend, will check this on Monday - let me know thoughts re 500s and I'll make changes if required. 

Thanks

thinkingserious commented:
Hi @dylan-asos,

Have a great weekend and thank you for all of the great progress!

Regarding the retry strategy for responses 500 to 504, we suggest a configurable back off strategy.

dylan-asos commented:
Think this is done now unless any further feedback, 

* Defining a set of 500 errors that should be retried 

* Introduced Min/Max/Delta concept to work out exponential backoff, as per https://github.com/Azure/elastic-db-tools/blob/master/Src/ElasticScale.Client/ElasticScale.Common/TransientFaultHandling/Implementation/ExponentialBackoff.cs

* Made reliability settings class immutable and verify settings in construct

* Split out tests for retry handler / settings classes

* Updated usage documentation, https://github.com/dylan-asos/sendgrid-csharp/blob/master/USE_CASES.md#transient_faults

Let me know if any issues / feedback / changes required. 

Cheers

thinkingserious commented:
Hello @dylan-asos,<br /><br/>Thanks again for the PR!<br />                                              <br />We appreciate your contribution and look forward to                                              continued collaboration. Thanks!                                             <br /><br />Team SendGrid DX

dylan-asos commented:
If any issues are raised just tag me in the thread & I'll dive in
